processor/processor.go

```diff
@@ -1875,6 +2007,124 @@ func (proc *HandleT) mainLoop(ctx context.Context) {
 	}
 }
 
+func (proc *HandleT) pipelineWithPause(ctx context.Context, fn func(ctx context.Context)) {
+	for {
+		var pause *PauseT
+
+		ctxPausable, cancel := context.WithCancel(ctx)
+		go func() {
+			pause = <-proc.pauseChannel
+			cancel()
+		}()
+		fn(ctxPausable)
+		// stop due to parent context cancellation
+		if ctx.Err() != nil {
+			return
+		}
+
+		// handle pause
+		if pause == nil {
+			panic("`pause` should not be nil")
+		}
+		proc.paused = true
+		pause.respChannel <- true
+		<-proc.resumeChannel
+	}
+}
+
+// mainPipeline: new way of handling jobs
+//
+// [getJobs] -chProc-> [processJobsForDest] -chTrans-> [transformations] -chStore-> [Store]
+func (proc *HandleT) mainPipeline(ctx context.Context) {
+	//waiting for reporting client setup
+	if proc.reporting != nil && proc.reportingEnabled {
+		proc.reporting.WaitForSetup(ctx, types.CORE_REPORTING_CLIENT)
+	}
+	wg := sync.WaitGroup{}
+	bufferSize := pipelineBufferedItems
+
+	chProc := make(chan []*jobsdb.JobT, bufferSize)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer close(chProc)
+
+		nextSleepTime := time.Duration(0)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(nextSleepTime):
+				if !isUnLocked {
+					nextSleepTime = proc.maxLoopSleep
+					continue
+				}
+
+				jobs := proc.getJobs()
+				if len(jobs) == 0 {
+					// no jobs found, double sleep time until maxLoopSleep
+					nextSleepTime = 2 * nextSleepTime
+					if nextSleepTime > proc.maxLoopSleep {
+						nextSleepTime = proc.maxLoopSleep
+					} else if nextSleepTime == 0 {
+						nextSleepTime = proc.readLoopSleep
+					}
+					continue
+				}
+
+				err := proc.markExecuting(jobs)
+				if err != nil {
+					pkgLogger.Error(err)
+					panic(err)
+				}
+
+				events := 0
+				for i := range jobs {
+					events += jobs[i].EventCount
+				}
+				// nextSleepTime is dependent on the number of events read in this loop
+				emptyRatio := 1.0 - math.Min(1, float64(events)/float64(maxEventsToProcess))
+				nextSleepTime = time.Duration(emptyRatio * float64(proc.readLoopSleep))
+
+				chProc <- jobs
+			}
+		}
+	}()
+
+	chTrans := make(chan transformationMessage, bufferSize)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer close(chTrans)
+		for jobs := range chProc {
+			chTrans <- proc.processJobsForDest(jobs, nil)
```

Comment:
We are always calling this function with second argument as nil?

Reply:
Yea that correct. My intention was to remove it in the next PR actually.